### ISSUE TITLE
Unbound variable aws_default_args

### ISSUE THREAD

**Issue by sj26** [buildkite-plugins/ecs-deploy-buildkite-plugin#91]
While trying to use this plugin I'm getting an error:

<img width="1417" alt="image" src="https://user-images.githubusercontent.com/14028/227090341-bb280c31-cafa-420f-b3ee-1f92c46159f1.png">

https://buildkite.com/sj26-test/buildkite-ecs-oidc-example/builds/28#01870c6d-495b-4747-9afb-2382a87f3c4a

```
~/.buildkite-agent/plugins/github-com-buildkite-plugins-ecs-deploy-buildkite-plugin/hooks/command: line 147: aws_default_args[@]: unbound variable
🚨 Error: The command exited with status 1
user command error: The plugin ecs-deploy command hook exited with status 1
```

Seems to be around here:

https://github.com/buildkite-plugins/ecs-deploy-buildkite-plugin/blob/master/hooks/command#L144-L148

I can see that it's defined earlier in the file, and so should be available:

https://github.com/buildkite-plugins/ecs-deploy-buildkite-plugin/blob/master/hooks/command#L35

My bash-fu is weak, though, so is this something weird to do with arrays? I remember something about bash 5 causing issues?

I'm running an agent on macOS:

```
bash --version
GNU bash, version 5.2.2(1)-release (aarch64-apple-darwin21.6.0)
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

**Comment by sj26** [comment]
I modified my plugin locally to use dash parameter expansion, but I'm still baffled as to why this happens:

https://fvue.nl/wiki/Bash:_Error_%60Unbound_variable%27_when_appending_to_empty_array#Solution_1:_Dash_.28-.29_parameter_expansion

**Comment by toote** [comment]
That is weird... we've seen this issue - and even corrected one instance of that (see #81) - but only when using Bash 4.3 as that particular version treats empty arrays as undefined variables. And yet you state that you are using bash 5.2.

If you define the `region` option of the plugin, does the issue go away?

**Comment by sj26** [comment]
If I define a region I get the same error but in a different place for a different variable:

```
Registered buildkite-ecs-oidc-example:11
~/.buildkite-agent/plugins/github-com-buildkite-plugins-ecs-deploy-buildkite-plugin/hooks/command: line 66: result[@]: unbound variable
🚨 Error: The command exited with status 1
user command error: The plugin ecs-deploy command hook exited with status 1
```

so I think bash is treating empty arrays as undefined.

I found another clue. I notice the hook has the shebang:

```
#/bin/bash
```

On my system (macOS 12.6.3) this is a different version of bash:

```
/bin/bash --version
GNU bash, version 3.2.57(1)-release (arm64-apple-darwin21)
Copyright (C) 2007 Free Software Foundation, Inc.
```

In that version of bash, an empty array is considered unset:

```
bash-3.2$ set -u
bash-3.2$ array=()
bash-3.2$ echo "${array[@]}"
bash: array[@]: unbound variable
```

It's explored here on stackoverflow, too:

https://stackoverflow.com/questions/7577052/bash-empty-array-expansion-with-set-u

I think to use arrays portably in this manner you will need to use the idiom:

```
${arr[@]+"${arr[@]}"}
```
